packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts

```diff
@@ -0,0 +1,209 @@
+  }
+
+  get iterationScopeType(): ScopeType | CustomScopeType {
+    return this.scopeHandler.iterationScopeType;
+  }
+
+  generateScopeCandidates(
+    editor: TextEditor,
+    position: Position,
+    direction: Direction,
+    _hints: ScopeIteratorRequirements,
+  ): Iterable<TargetScope> {
+    return direction === "backward"
+      ? this.generateScopeCandidatesBackward(editor, position)
+      : this.generateScopeCandidatesForward(editor, position);
+  }
+
+  private *generateScopeCandidatesBackward(
+    editor: TextEditor,
+    position: Position,
+  ): Iterable<TargetScope> {
+    let targetRangeForward = next(
+      generateTargetRangesInDirection(
+        this.scopeHandler,
+        editor,
+        position,
+        "forward",
+      ),
+    );
+
+    const targetRangesBackwardIter = generateTargetRangesInDirection(
+      this.scopeHandler,
+      editor,
+      position,
+      "backward",
+    );
+
+    for (const targetRange of targetRangesBackwardIter) {
+      if (
+        targetRangeForward != null &&
+        isAdjacent(targetRange[1], targetRangeForward[0])
+      ) {
+        yield targetsToScope(targetRange[0], targetRangeForward[1]);
+        targetRangeForward = undefined;
+      } else {
+        yield targetsToScope(...targetRange);
+      }
+    }
+  }
+
+  private *generateScopeCandidatesForward(
+    editor: TextEditor,
+    position: Position,
+  ): Iterable<TargetScope> {
+    let targetRangeBackward = next(
+      generateTargetRangesInDirection(
+        this.scopeHandler,
+        editor,
+        position,
+        "backward",
+      ),
+    );
+
+    const targetRangesForwardIter = generateTargetRangesInDirection(
+      this.scopeHandler,
+      editor,
+      position,
+      "forward",
+    );
+
+    for (const targetRange of targetRangesForwardIter) {
+      if (
+        targetRangeBackward != null &&
+        isAdjacent(targetRangeBackward[1], targetRange[0])
+      ) {
+        yield targetsToScope(targetRangeBackward[0], targetRange[1]);
+        targetRangeBackward = undefined;
+      } else {
+        yield targetsToScope(...targetRange);
+      }
+    }
+  }
+}
+
+function targetsToScope(
+  leadingTarget: Target,
+  trailingTarget: Target,
+): TargetScope {
+  if (leadingTarget.contentRange.isRangeEqual(trailingTarget.contentRange)) {
+    return {
+      editor: leadingTarget.editor,
+      domain: leadingTarget.contentRange,
+      getTargets: () => [leadingTarget],
+    };
+  }
+
+  const range = leadingTarget.contentRange.union(trailingTarget.contentRange);
+  return {
+    editor: leadingTarget.editor,
+    domain: range,
+    getTargets: () => [leadingTarget.withContentRange(range)],
+  };
+}
+
+function* generateTargetRangesInDirection(
+  scopeHandler: ScopeHandler,
+  editor: TextEditor,
+  position: Position,
+  direction: Direction,
+): Iterable<[Target, Target]> {
+  const isForward = direction === "forward";
+  let first, last: Target | undefined;
```

Comment:
are these in iteration order or document order? We really need to have solid conventions when we could be iterating backwards

Reply:
First and last is in document order. The pair is yielded in iteration order.

Reply:
Ok I'd call them `proximal` and `distal` then. That's the convention I've used elsewhere. "first" and "last" may be expected to correspond to the modifiers of those names, which are document order